- Added spin state order handling to PolarizationCorrectionFredrikze Correction Algorithm

yusufjimoh opened the pull request:
### Description of work
- Implemented InputSpinStateOrder and OutputSpinStateOrder properties.
- Updated execPA functions to dynamically handle different spin state orders.
- Extracted input and output spin state mapping logic into separate functions for better code organization and reusability.
- Ensured backward compatibility by setting default spin state orders.
- Added validation to ensure input workspace count matches the spin state entered
- Updated PolarizationEfficiencyCor algorithm to handle spin states for the PolarizationCorrectionFredrikze algorithm
- Updated Documentations

#### Purpose of work
#36981 

### To test:
- Execute the PolarizationCorrectionFredrikze Correction Algorithm
- Test with Input SpinState and/or Output Spin State, algorithm should run successfully
- Test without input SpinState and/or Output Spin State, confirm it still executes successfully
- Execute the PolarizationEfficiencyCor algorithm, running the PolarizationCorrectionFredrikze algorithm as a child. The algorithm should run successfully

yusufjimoh committed:
- Added spin state order handling to PolarizationCorrectionFredrikze algorithm

- Implemented InputSpinStateOrder and OutputSpinStateOrder properties.
- Updated execPA functions to dynamically handle different spin state orders.
- Extracted input and output spin state mapping logic into separate functions for better code organization and reusability.
- Ensured backward compatibility by setting default spin state orders.
- Added validation to ensure input workspace count matches the spin state entered

yusufjimoh committed:
- Added spin state order handling to PolarizationCorrectionFredrikze algorithm

- Implemented InputSpinStateOrder and OutputSpinStateOrder properties.
- Updated execPA functions to dynamically handle different spin state orders.
- Extracted input and output spin state mapping logic into separate functions for better code organization and reusability.
- Ensured backward compatibility by setting default spin state orders.
- Added validation to ensure input workspace count matches the spin state entered

yusufjimoh committed:
- Included unit tests to verify correct functionality and validation of spin state orders.

yusufjimoh committed:
- splitted the spin state order string to a string of vectors using the polarization correction helper splitSpinStateString

yusufjimoh committed:
added p and a to spin states

yusufjimoh committed:
- modified unit tests to test spin state orders

yusufjimoh committed:
- updated the documentation to describe the spin states and its usage

sf1919 commented:
This appears to have cppcheck failures which are unrelated to recent isues with the Mantid pipelines.

yusufjimoh committed:
Add Input and Output SpinStates for FredrikzeAlgorithm in PolarizationEfficiencyCor

- added input and output spin states for fredrikze algorithm in polarizationefficiencycor algorithm
- updated unit tests to test spin states in both algorithms
- updated polarizationefficiencycor documentation
- added a release note for the update

cailafinn commented on Framework/Algorithms/src/PolarizationCorrectionFredrikze.cpp:
```diff
@@ -247,10 +327,17 @@ WorkspaceGroup_sptr PolarizationCorrectionFredrikze::execPA(const WorkspaceGroup
+                                                             const std::vector<std::string> &outputOrder) {
+
+  // Default order for PNR analysis
+  std::vector<std::string> defaultOrder = {SPIN_STATES::P, SPIN_STATES::A};
```
Similarly with this one

cailafinn commented on Framework/Algorithms/src/PolarizationCorrectionFredrikze.cpp:
```diff
@@ -247,10 +327,17 @@ WorkspaceGroup_sptr PolarizationCorrectionFredrikze::execPA(const WorkspaceGroup
+  // Default order for PNR analysis
+  std::vector<std::string> defaultOrder = {SPIN_STATES::P, SPIN_STATES::A};
+  // Map the input workspaces according to the specified input order
+  auto inputMap = mapOrderToWorkspaces(inWS, inputOrder, defaultOrder);
```
Could this be a `const &` too? I don't think it gets manipulated from here out.

cailafinn commented on Framework/Algorithms/src/PolarizationEfficiencyCor.cpp:
```diff
@@ -26,13 +27,16 @@ namespace {
 namespace Prop {
 static const std::string FLIPPERS{"Flippers"};
-static const std::string SPIN_STATES{"SpinStates"};
+static const std::string SPIN_STATES{"SpinStatesOutWildes"};
```
We should rename this variable to clearly differenciate it from the Fredrikze ones.
```suggestion
static const std::string OUTPUT_WILDES_SPIN_STATES{"SpinStatesOutWildes"};
```

cailafinn commented on Framework/Algorithms/src/PolarizationEfficiencyCor.cpp:
```diff
@@ -117,9 +121,38 @@ void PolarizationEfficiencyCor::init() {
                   "(Fredrikze method only)");
 
+  const auto fredrikzeSpinStateValidator =
+      std::make_shared<SpinStateValidator>(std::unordered_set<int>{2, 4}, true, 'p', 'a', true);
```
We should make use of the constants here if we can.

cailafinn commented on Framework/Algorithms/test/PolarizationCorrectionFredrikzeTest.h:
```diff
@@ -362,4 +362,125 @@ class PolarizationCorrectionFredrikzeTest : public CxxTest::TestSuite {
+
+  void test_valid_spin_state_order_for_PA() {
+    auto groupWS = std::make_shared<WorkspaceGroup>();
+    groupWS->addWorkspace(create1DWorkspace(4, 1, 1));
+    groupWS->addWorkspace(create1DWorkspace(4, 1, 1));
+    groupWS->addWorkspace(create1DWorkspace(4, 1, 1));
+    groupWS->addWorkspace(create1DWorkspace(4, 1, 1));
+
+    PolarizationCorrectionFredrikze alg;
+    alg.setChild(true);
+    alg.setRethrows(true);
+    alg.initialize();
+    alg.setProperty("InputWorkspace", groupWS);
+    alg.setPropertyValue("OutputWorkspace", "dummy");
+    alg.setProperty("PolarizationAnalysis", "PA");
```
This group workspace creation and algorithm initialisation could probably be consolidated into their own methods, they're repeated a few times in the new tests. You could have an if statement in there for producing the smaller PNR ones as well.

cailafinn commented on docs/source/algorithms/PolarizationCorrectionFredrikze-v1.rst:
```diff
@@ -1,3 +1,7 @@
+To update your documentation based on the new approach of using spin states directly rather than flipper configurations, you need to revise the relevant sections to reflect this change. Here’s an updated version of your documentation with a focus on the new spin state configurations and removing references to flipper configurations:
+
+---
```
We should highlight things like this in the release notes rather than at the top of the docs, which have a standardised style.
```suggestion
```

cailafinn commented on docs/source/algorithms/PolarizationCorrectionFredrikze-v1.rst:
```diff
@@ -9,28 +13,43 @@
 Performs wavelength polarization correction on a TOF reflectometer spectrometer.
 
-Algorithm is based on the paper Fredrikze, H, et al. "Calibration of a polarized neutron reflectometer" Physica B 297 (2001).
+The algorithm is based on the paper Fredrikze, H, et al. "Calibration of a polarized neutron reflectometer" Physica B 297 (2001).
```
Could we add an explicit reference here? (The doc for `DepolarizedAnalyserTransmission.rst` has an example of how to do this.

cailafinn commented on docs/source/algorithms/PolarizationCorrectionFredrikze-v1.rst:
```diff
@@ -9,28 +13,43 @@
-Since this ratio is wavelength dependent, alpha is a polynomial, which is expressed as a function of wavelength. For example:
-:math:`\alpha(\lambda) =\sum\limits_{i=0}^{i=2} K_{i}\centerdot\lambda^i`, can be provided as :math:`K_{0}, K_{1}, K_{2}`
+- :literal:`'pp, pa, ap, aa'`
+   Full polarization corrections: both parallel, parallel-anti-parallel, anti-parallel-parallel, both  anti-parallel. Four input workspaces are required. The spin states can be provided in any order and should match the order of the workspaces in the input group.
+
+- :literal:`'p, a'`
+   PNR polarization corrections:  parallel, anti-parallel. Two input workspaces are required. The spin states can be provided in any order and should match the order of the workspaces in the input group.
```
Can we make clear here that `PA` is Full polarization corrections and what `PNR` stands for?

cailafinn commented on docs/source/algorithms/PolarizationCorrectionFredrikze-v1.rst:
```diff
@@ -9,28 +13,43 @@
-In the ideal case :math:`P_{p} = P_{a} = A_{p} = A_{a} = 1`
+Spin State Configurations
+#########################
+The algorithm directly uses spin states to define the measurement setup. The spin states are specified using two properties:
```
```suggestion
The algorithm expresses the order of the workspaces in the input and output groups in terms of spin states. These orders are specified using two properties:
```

cailafinn commented on docs/source/algorithms/PolarizationCorrectionFredrikze-v1.rst:
```diff
@@ -9,28 +13,43 @@
+**Output Spin State Order:**
+This property determines the order of the spin states in the output workspace group. Similar to the input configuration, users can specify output spin states in any order.
 
 Output from Full Polarization Analysis
```
Can we have an extra heading for the outputs from the PNR version?

cailafinn commented on docs/source/algorithms/PolarizationCorrectionFredrikze-v1.rst:
```diff
@@ -41,7 +60,11 @@ Entry in group  Measurement
 .. categories::
 
 .. sourcelink::
+
+---
+
```
Is there a reason for this extra line being included?

cailafinn reviewed (changes requested):
This is half a review as I didn't have time to do the functional testing, but I had a few code comments to mention from that half:

cailafinn reviewed (changes requested):
Tested functionally with the following script. Changes are working well for the input and output spin state orders when the right number of characters are used, setting both properly.
```python
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

run = Load("POLREF00004699.nxs")
eff = Load("fredrikze_efficiencies_37125.nxs")
run = ConvertUnits(run, "Wavelength")
eff = ConvertUnits(eff, "Wavelength")
eff = RebinToWorkspace(eff, run.getItem(0), True)
out = PolarizationCorrectionFredrikze(run, eff, "PNR", OutputSpinStateOrder="p, a")
```
When messing with the script, found a couple of other issues:
Setting the `OutputSpinStateOrder` to "pp, aa" ran the algorithm and threw an `IndexError: map::at:  key not found`, this should be checked in the `validateInputs()` method. 
Setting the `InputSpinStateOrder` to "pp, aa" caused a segfault that crashed to desktop.

```python
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

run1 = Load("POLREF00004699.nxs")
run2 = Load("POLREF00004699.nxs")
run2 = run2 * 1.2
run = GroupWorkspaces([run1, run2])

eff = Load("fredrikze_efficiencies_37125.nxs")

run = ConvertUnits(run, "Wavelength")
eff = ConvertUnits(eff, "Wavelength")

eff = RebinToWorkspace(eff, run.getItem(0), True)

out = PolarizationCorrectionFredrikze(run, eff, "PA", OutputSpinStateOrder="pp, aa, pa, ap")
```
The inverse applies to the above, using `p` and `a` in this one.

yusufjimoh committed:
Refactor Fredrikze Algorithm SpinStates Implementation

- Renamed the input and output spin states
- refactored tests to eliminate code duplication
- added an helper function to fix crash which happens when user enter pp, aa as spinstates
- improved code based on PR review suggestions
- updated unit tests to handle crash which happens when user enter pp, aa as spinstates
- updated documentation

yusufjimoh commented:
Thank you for the comprehensive review. I have implemented the suggested changes and updated the unit tests to resolve the segfault crash

rbauststfc commented:
@yusufjimoh, thanks for all the hard work on this one. It's a change that can definitely wait until the next release, that's no problem, so I'm going to switch the milestone to 6.12 at this point. It will mean that the release notes will need to be moved into a different folder at an appropriate stage (we can't do that just yet), so I'll discuss that with you at some point in the future (no need to worry about it right now). Hope this sounds OK.

cailafinn commented on Framework/Algorithms/src/PolarizationCorrectionFredrikze.cpp:
```diff
@@ -22,26 +24,37 @@ using namespace Mantid::API;
+static const std::string OUTPUT_WORKSPACE{"OutputWorkspace"};
+static const std::string INPUT_SPIN_STATES("InputSpinStates");
+static const std::string OUTPUT_SPIN_STATES("OutputSpinStates");
+static const std::string crhoLabel("Rho");
```
I don't this this and some of the others in here (`cApLabel`etc.) are Properties of the algorithm, and so should probably be in a different namespace.

cailafinn commented on docs/source/algorithms/PolarizationCorrectionFredrikze-v1.rst:
```diff
@@ -41,7 +57,67 @@ Entry in group  Measurement
+Output from Polarized Neutron Reflectivity
+------------------------------------------
+
+The output of this algorithm is a :ref:`WorkspaceGroup <WorkspaceGroup>`, the resulting :ref:`WorkspaceGroup <WorkspaceGroup>` will have 2 entries, corresponding to the output specified spin state orders, if no specific output spin states are specified, it will be formatted as follows:
```
```suggestion
The output of this algorithm is a :ref:`WorkspaceGroup <WorkspaceGroup>`. The resulting :ref:`WorkspaceGroup <WorkspaceGroup>` will have 2 entries, corresponding to the specified output spin state orders. If no specific output spin states are specified, it will be formatted as follows:
```

cailafinn commented on docs/source/algorithms/PolarizationCorrectionFredrikze-v1.rst:
```diff
@@ -41,7 +57,67 @@ Entry in group  Measurement
+
+**Example - PolarizationCorrectionFredrikze (PNR)**
+
+.. testcode:: PolarizationCorrectionFredrikze (PNR)
+
+    # import mantid algorithms, numpy and matplotlib
+    from mantid.simpleapi import *
+    import matplotlib.pyplot as plt
+    import numpy as np
+
+    run = Load("POLREF00004699.nxs")
+    eff = Load("fredrikze_efficiencies_37125.nxs")
+    run = ConvertUnits(run, "Wavelength")
+    eff = ConvertUnits(eff, "Wavelength")
+    eff = RebinToWorkspace(eff, run.getItem(0), True)
+    out = PolarizationCorrectionFredrikze(run, eff, "PNR", InputSpinStates="pp, pa, ap, aa", OutputSpinStates="p, a")
+
```
There should probable be a `.. testoutput::` here as in the one from the wildes docs. That way we can make sure this usage example consistently performs correctly.

cailafinn commented on docs/source/algorithms/PolarizationCorrectionFredrikze-v1.rst:
```diff
@@ -41,7 +57,67 @@ Entry in group  Measurement
+    run = ConvertUnits(run, "Wavelength")
+    eff = ConvertUnits(eff, "Wavelength")
+    eff = RebinToWorkspace(eff, run.getItem(0), True)
+    out = PolarizationCorrectionFredrikze(run, eff, "PNR", InputSpinStates="pp, pa, ap, aa", OutputSpinStates="p, a")
```
Does a PNR example typically take 4 workspaces an input or should it only take two?

cailafinn reviewed (dismissed):
Functionally tested using previous scripts and working well. Small issue with validation mentioned in code comments. Just a few more questions about the code, but this is looking good:

rbauststfc commented on Framework/Algorithms/src/PolarizationEfficiencyCor.cpp:
```diff
@@ -210,7 +258,7 @@ void PolarizationEfficiencyCor::checkFredrikzeProperties() const {
   }
-  if (!isDefault(Prop::SPIN_STATES)) {
+  if (!isDefault(Prop::OUTPUT_WILDES_SPIN_STATES)) {
     throw std::invalid_argument("Property SpinStates cannot be used with the Fredrikze method.");
```
I think this error message needs to say "Property SpinStatesOutWildes cannot be used with the Fredrikze method." now that the Wildes spin states property has been renamed.

rbauststfc reviewed (commented):
Looks like a lot of tidy-ups have gone into this PR, thanks very much for all of your hard work on this Yusuf!

I haven't completed functional testing yet, but I've checked the validation, which all seems very robust.

I've noticed a few places in the documentation and some of the validation messages where it might be good to clarify some of the wording, or where there are typos, so I've added relevant comments to look at those final bits, if that's OK.

I think that there are one or two places where we might like to look at our use of constants in this algorithm (mostly in the existing rather than new code), but given that quite a bit of refactoring and improvements have already been made in this PR, I will open a separate maintenance issue to look at these last bits.

Additionally, we've added unit tests that provide good coverage of the input and output spin state validation, which is great for now. As a follow-up maintenance task, I think it would be good to think about how we might test that the algorithm is applying the re-ordering correctly, as I think we're currently just assuming that if the algorithm doesn't throw an error then it's all worked OK. I will add a separate issue to look at this, as I don't want to hold up this PR too much longer.

rbauststfc commented:
Thanks for the documentation updates Yusuf, those are all looking good.

Just to flag up, the docs build failure on this PR is genuine as the example code gives an error when run in workbench. I think the following should hopefully work OK. I would hold off updating this while I finish my functional testing, though, just to make sure this will be the last change. I will let you know when I've finished that.

For PNR:

```
run = Load("POLREF00004699.nxs")
ws1 = CreateWorkspace([0.01, 17.0, 34.0], [50000, 50000])
eff = JoinISISPolarizationEfficiencies(Pp=ws1,   Rho=ws1, Ap=ws1, Alpha=ws1)
run = ConvertUnits(run, "Wavelength")
eff = ConvertUnits(eff, "Wavelength")
eff = RebinToWorkspace(eff, run.getItem(0), True)

out = PolarizationCorrectionFredrikze(run, eff, "PNR", InputSpinStates="p, a", OutputSpinStates="p, a")

for i, orig in enumerate(run):
	corrected = out[i]
	y_idx = run[0].yIndexOfX(15.0)
	ratio = corrected.readY(29)[y_idx] / orig.readY(29)[y_idx]
	print(f"Ratio of corrected {corrected.name()} and original {orig.name()} intensity at 15A: {round(ratio, 4)}")
```

Which gives output:

```
   Ratio of corrected out_1 and original run_1 intensity at 15A: 1.0005
   Ratio of corrected out_2 and original run_2 intensity at 15A: 0.9508
```


For PA:

```
run1 = Load("POLREF00004699.nxs")
run2 = run1 * 1.2
run = GroupWorkspaces([run1, run2])
ws1 = CreateWorkspace([0.01, 17.0, 34.0], [50000, 50000])
eff = JoinISISPolarizationEfficiencies(Pp=ws1,   Rho=ws1, Ap=ws1, Alpha=ws1)
run = ConvertUnits(run, "Wavelength")
eff = ConvertUnits(eff, "Wavelength")
eff = RebinToWorkspace(eff, run.getItem(0), True)

out = PolarizationCorrectionFredrikze(run, eff, "PA", InputSpinStates="pp, pa, ap, aa", OutputSpinStates="pp, pa, ap, aa")

for i, orig in enumerate(run):
	corrected = out[i]
	y_idx = run[0].yIndexOfX(15.0)
	ratio = corrected.readY(29)[y_idx] / orig.readY(29)[y_idx]
	print(f"Ratio of corrected {corrected.name()} and original {orig.name()} intensity at 15A: {round(ratio, 4)}")
```

Which gives output:

```
   Ratio of corrected out_1 and original run1_1 intensity at 15A: 1.0025
   Ratio of corrected out_2 and original run1_2 intensity at 15A: 0.9527
   Ratio of corrected out_3 and original run2_1 intensity at 15A: 0.8355
   Ratio of corrected out_4 and original run2_2 intensity at 15A: 0.794
```

rbauststfc commented:
Functional testing looks really good to me.👍  If we can just replace the example in the docs with the above then hopefully that should be the last thing left before this can go in.

yusufjimoh commented:
> Functional testing looks really good to me.👍 If we can just replace the example in the docs with the above then hopefully that should be the last thing left before this can go in.



> Thanks for the documentation updates Yusuf, those are all looking good.
> 
> Just to flag up, the docs build failure on this PR is genuine as the example code gives an error when run in workbench. I think the following should hopefully work OK. I would hold off updating this while I finish my functional testing, though, just to make sure this will be the last change. I will let you know when I've finished that.
> 
> For PNR:
> 
> ```
> run = Load("POLREF00004699.nxs")
> ws1 = CreateWorkspace([0.01, 17.0, 34.0], [50000, 50000])
> eff = JoinISISPolarizationEfficiencies(Pp=ws1,   Rho=ws1, Ap=ws1, Alpha=ws1)
> run = ConvertUnits(run, "Wavelength")
> eff = ConvertUnits(eff, "Wavelength")
> eff = RebinToWorkspace(eff, run.getItem(0), True)
> 
> out = PolarizationCorrectionFredrikze(run, eff, "PNR", InputSpinStates="p, a", OutputSpinStates="p, a")
> 
> for i, orig in enumerate(run):
> 	corrected = out[i]
> 	y_idx = run[0].yIndexOfX(15.0)
> 	ratio = corrected.readY(29)[y_idx] / orig.readY(29)[y_idx]
> 	print(f"Ratio of corrected {corrected.name()} and original {orig.name()} intensity at 15A: {round(ratio, 4)}")
> ```
> 
> Which gives output:
> 
> ```
>    Ratio of corrected out_1 and original run_1 intensity at 15A: 1.0005
>    Ratio of corrected out_2 and original run_2 intensity at 15A: 0.9508
> ```
> 
> For PA:
> 
> ```
> run1 = Load("POLREF00004699.nxs")
> run2 = run1 * 1.2
> run = GroupWorkspaces([run1, run2])
> ws1 = CreateWorkspace([0.01, 17.0, 34.0], [50000, 50000])
> eff = JoinISISPolarizationEfficiencies(Pp=ws1,   Rho=ws1, Ap=ws1, Alpha=ws1)
> run = ConvertUnits(run, "Wavelength")
> eff = ConvertUnits(eff, "Wavelength")
> eff = RebinToWorkspace(eff, run.getItem(0), True)
> 
> out = PolarizationCorrectionFredrikze(run, eff, "PA", InputSpinStates="pp, pa, ap, aa", OutputSpinStates="pp, pa, ap, aa")
> 
> for i, orig in enumerate(run):
> 	corrected = out[i]
> 	y_idx = run[0].yIndexOfX(15.0)
> 	ratio = corrected.readY(29)[y_idx] / orig.readY(29)[y_idx]
> 	print(f"Ratio of corrected {corrected.name()} and original {orig.name()} intensity at 15A: {round(ratio, 4)}")
> ```
> 
> Which gives output:
> 
> ```
>    Ratio of corrected out_1 and original run1_1 intensity at 15A: 1.0025
>    Ratio of corrected out_2 and original run1_2 intensity at 15A: 0.9527
>    Ratio of corrected out_3 and original run2_1 intensity at 15A: 0.8355
>    Ratio of corrected out_4 and original run2_2 intensity at 15A: 0.794
> ```

I have updated the documentation with the above code, Thanks.

yusufjimoh committed:
--amend-no-edit

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

rbauststfc reviewed (approved):
Thanks very much for all your hard work on this issue Yusuf.

The code is working as expected and provides the additional functionality we needed under the original issue. There has also been a fair bit of maintenance pulled into this PR like some minor refactoring and adding usage examples to the docs, which are all looking good. There is some additional maintenance we could do and some unit test improvements that I think we could make, but it would be good to look at these under separate issues so that we can get this in and allow other related work to progress  (see #38287 and #38288).
